[HttpFoundation] Add mime type guessing by file extension

ipf opened the pull request:
This adds mime type guessing by file extension.
Sometimes file information is only stored as metadata, so we don't have access
to a file. So, for guessing the mime type this does it just by using a string (i.e. 'jpg).

| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | no
| License       | MIT

Hanmac commented:
i am okay with adding such a "extension" to "mime-type" guesser,
but i would prefer if it can use some system files like that, like the mime.types from apache.

xabbuh commented on src/Symfony/Component/HttpFoundation/File/MimeType/MimeTypeByExtensionGuesser.php:
```diff
@@ -0,0 +1,1019 @@
+     *
+     * @var array
+     */
+    protected $defaultExtensions = array(
```
`private`

xabbuh commented on src/Symfony/Component/HttpFoundation/File/MimeType/MimeTypeByExtensionGuesser.php:
```diff
@@ -0,0 +1,1019 @@
+    /**
+     * {@inheritdoc}
+     */
+    public function guess($extension)
```
This is wrong. The interface states that `guess()` must expect the file path and not only its extension.

xabbuh commented:
What about registering this also as an alternative in `MimeTypeGuesser` that will be used when neither `FileinfoMimeTypeGuesser` nor `FileBinaryMimeTypeGuesser` yield a result?

xabbuh commented:
Alternatively, instead of inlining the map of extensions to mime types inside the class, we could include the original resources file and allow to pass the file to be used in the constructor (defaulting to the file that is shipped with the HttpFoundation component).

ipf committed:
[HttpFoundation] Add mime type guessing by file extension

This adds mime type guessing by file extension.
Sometimes file information is only stored as metadata, so we don't
have access to a file. So, for guessing the mime type this does it just
by using a string (i.e. 'jpg).

| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | no
| License       | MIT

ipf commented:
Hi @xabbuh, thanks a lot for your reviews. Rebuilding the `MimeTypeGuesser` is definetly a nice thing, but rebuilding all that is quite a lot of work.
Using the original file is also a great idea, but parsing it on each call may be too much overhead instead of my small and simple solution. What do you think?

xabbuh commented:
> but rebuilding all that is quite a lot of work

What do you mean with "rebuilding all that"?

> Using the original file is also a great idea, but parsing it on each call may be too much overhead instead of my small and simple solution.

This should indeed not happen inside the constructor to avoid overhead for requests that don't make use of the mime type guesser or for cases where one of the other guesser returned a result, but only when the extension based guesser actually does have to perform some work. I don't expect the parsing to be too heavy in these cases. But you may prove me wrong and we should then indeed think about some caching.

dosten commented on src/Symfony/Component/HttpFoundation/File/MimeType/MimeTypeByExtensionGuesser.php:
```diff
@@ -0,0 +1,1023 @@
+     */
+    public function guess($extension)
+    {
+        return isset($this->defaultExtensions[$extension]) ? $this->defaultExtensions[$extension] : null;
```
I don't know if it worth it, but PNG, JPEG, etc are valid extensions, what about lowercase the extension before use it?

dmaicher commented on src/Symfony/Component/HttpFoundation/Tests/File/MimeType/MimeTypeByExtensionGuesserTest.php:
```diff
@@ -0,0 +1,56 @@
+     */
+    public function testMimeTypeGuessing($extension, $mimeType)
+    {
+        $result = $this->guesser->guess($extension);
```
could be a bit shorter like this? Imho no setup method or private member needed here?

```php
$result = (new MimeTypeByExtensionGuesser())->guess($extension);
```

dmaicher commented on src/Symfony/Component/HttpFoundation/Tests/File/MimeType/MimeTypeByExtensionGuesserTest.php:
```diff
@@ -0,0 +1,56 @@
+    /**
+     * @dataProvider extensionDataProvider
+     *
+     * @param $extension
```
useless without any type info

nicolas-grekas commented:
I agree with @xabbuh about the list: we dont want to maintain it, so we definitely should build it from a reference source. This could happen with a cache warmer I guess.

nicolas-grekas commented:
Moving to 4.1. Rebase on master might be needed, where PHP 7.1 features can be used btw.

fabpot commented:
The list of extensions is already part of the `MimeTypeExtensionGuesser` class, so I would extract it from there so that we can reuse it easily. Anyway, any interest in finishing this one? Or should it be closed?

teohhanhui commented:
This may not be the right place to raise the issue, but the DX around MIME type guessing is kind of broken. See https://github.com/symfony/symfony/issues/15460#issuecomment-380138425 for example.

fabpot commented:
Closing in favor of #29896
